src/OpenLoco/Ui/ViewportInteraction.cpp

```diff
@@ -578,4 +578,80 @@ namespace OpenLoco::Ui::ViewportInteraction
         }
         return { Pos2(mapPos.x & 0xFFE0, mapPos.y & 0xFFE0) };
     }
+
+    // 0x0045F1A7
+    std::optional<std::pair<Pos2, Ui::Viewport*>> getSurfaceLocFromUi(const xy32& screenCoords)
+    {
+        auto [info, viewport] = Ui::ViewportInteraction::getMapCoordinatesFromPos(screenCoords.x, screenCoords.y, ~Ui::ViewportInteraction::InteractionItemFlags::surface);
+
+        if (info.type == Ui::ViewportInteraction::InteractionItem::noInteraction)
+        {
+            return {};
+        }
+
+        const auto minPosition = info.pos;                  // E40128/A
+        const auto maxPosition = info.pos + Pos2{ 31, 31 }; // E4012C/E
+        auto mapPos = info.pos + Pos2{ 16, 16 };
+        const auto initialVPPos = viewport->uiToMap(screenCoords);
+
+        for (int32_t i = 0; i < 5; i++)
+        {
+            const auto z = TileManager::getHeight(mapPos);
+            mapPos = Ui::viewportCoordToMapCoord(initialVPPos.x, initialVPPos.y, z, viewport->getRotation());
+            mapPos.x = std::clamp(mapPos.x, minPosition.x, maxPosition.x);
+            mapPos.y = std::clamp(mapPos.y, minPosition.y, maxPosition.y);
+        }
+
+        return { std::make_pair(mapPos, viewport) };
+    }
+
+    // 0x0045FE05
+    // NOTE: Original call getSurfaceLocFromUi within this function
+    // instead OpenLoco has split it in two. Also note that result of original
+    // was a Pos2 start i.e. (& 0xFFE0) both components
+    static uint8_t getQuadrantFromPos(const Map::Pos2& loc)
+    {
+        const auto xNibble = loc.x & 0x1F;
+        const auto yNibble = loc.y & 0x1F;
+        if (xNibble > 16)
+        {
+            if (yNibble >= 16)
+            {
+                return 0;
+            }
+            else
+            {
+                return 1;
+            }
```

Comment:
This seems like an ideal case for using a ternary operator. What do you think?
```suggestion
            return yNibble >= 16 ? 0 : 1;
```